migrations/2025-04-25-090000_trusted-publishing/up.sql

```diff
@@ -0,0 +1,64 @@
+create table github_oidc_configs
```

Comment:
does it make sense to have a table per OIDC provider? or would it be better to save them all in the same table?

since they all use different claims I lean towards table per provider, but I'm open to the opposite if there are good reasons for it.

Reply:
I lean towards starting with a single table for two reasons: 1. It's the simplest way to begin. 2. After a quick glance at the given columns, I don't anticipate significant variations between different providers, and it seems like a good starting point for all providers already.

Reply:
> I don't anticipate significant variations between different providers

PyPI started to support Google as an OIDC provider for Trusted Publishing, which only uses two fields: email address (required) and subject (optional).

in other words, this is significantly different from what GitHub/GitLab support.

Reply:
Ah, good to know! I was only considering the repo holder previously :sweat_smile:. Then this indeed makes more sense to me, as it might vary for each provider!

I haven't investigated this yet, but if those columns won't be used for filtering and mainly serve as data, or if we only operate on them within the app, another possible solution could be to save those configurations as JSONB to accommodate different providers. Though the flow chart doesn't include this, I see an index clause (`github_oidc_configs (repository_owner_id, repository_name)`) in the schema, which means we might need these as filtering columns rather than just relying on either `id` or `crate_id`, so this might not be suitable?

Reply:
> which means we might need these as filtering columns

the way it works is that we receive a JWT with various "claims". the `iss` claim (issuer) tells us where the JWT is coming from (e.g. GitHub Actions), which lets us select the correct table to search for corresponding Trusted Publishing configurations. the rest of the provider-specific claims are then used to find one or more matching configurations (multiple crates could be in the same repository). from these configurations we then have a list of crate IDs, that the newly generated temporary access token is valid for.

so yes, these columns are mainly used for filtering to find matching configurations for a JWT.

Reply:
My initial reaction here was the same as @eth3lbert's, but I think I'm convinced that provider-specific tables are the way to go here.